File: iosApp/iosApp/Features/Chat/ConversationSettingsViewModel.swift
//
//  ConversationSettingsViewModel.swift
//  BDE
//
//  Created by Nathan Fallet on 07/07/2023.
//  Copyright © 2023 orgName. All rights reserved.
//

import Foundation
import shared

class ConversationSettingsViewModel: ObservableObject {
    
    let conversation: ChatConversation
    
    @Published var notifications: Bool
    @Published var members = [User]()
    
    init(conversation: ChatConversation) {
        self.conversation = conversation
        self.notifications = conversation.membership?.notifications ?? true
    }
    
    func onAppear(token: String?) {
        AnalyticsService.shared.log(.screenView(screenName: "conversation_settings", screenClass: "ConversationSettingsView"))
        
        fetchMembers(token: token)
    }
    
    func fetchMembers(token: String?) {
        guard let token else {
            return
        }
        Task {
            let members = try await CacheService.shared.apiService().getChatMembers(
                token: token,
                type: conversation.groupType,
                id: conversation.groupId
            )
            DispatchQueue.main.async {
                self.members = members
            }
        }
    }
    
    func updateMembership(token: String?) {
        guard let token else {
            return
        }
        Task {
            try await CacheService.shared.apiService().putChatMembers(
                token: token,
                type: conversation.groupType,
                id: conversation.groupId,
                upload: ChatMembershipUpload(
                    notifications: KotlinBoolean(bool: notifications)
                )
            )
        }
    }
    
}

File: androidApp/src/main/java/me/nathanfallet/bdeensisa/features/chat/ConversationSettingsViewModel.kt
package me.nathanfallet.bdeensisa.features.chat

import android.app.Application
import androidx.lifecycle.AndroidViewModel
import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.viewModelScope
import com.google.firebase.analytics.FirebaseAnalytics
import com.google.firebase.analytics.ktx.analytics
import com.google.firebase.analytics.ktx.logEvent
import com.google.firebase.ktx.Firebase
import kotlinx.coroutines.launch
import me.nathanfallet.bdeensisa.database.DatabaseDriverFactory
import me.nathanfallet.bdeensisa.extensions.SharedCacheService
import me.nathanfallet.bdeensisa.models.ChatConversation
import me.nathanfallet.bdeensisa.models.ChatMembershipUpload
import me.nathanfallet.bdeensisa.models.User

class ConversationSettingsViewModel(
    application: Application,
    token: String?,
    val conversation: ChatConversation
) : AndroidViewModel(application) {

    // Properties

    private val notifications = MutableLiveData<Boolean>()
    private val members = MutableLiveData<List<User>>()

    // Getters

    fun getNotifications(): LiveData<Boolean> {
        return notifications
    }

    fun getMembers(): LiveData<List<User>> {
        return members
    }

    // Setters

    fun setNotifications(value: Boolean) {
        notifications.value = value
    }

    // Methods

    init {
        Firebase.analytics.logEvent(FirebaseAnalytics.Event.SCREEN_VIEW) {
            param(FirebaseAnalytics.Param.SCREEN_NAME, "conversation_settings")
            param(FirebaseAnalytics.Param.SCREEN_CLASS, "ConversationSettingsView")
        }

        notifications.value = conversation.membership?.notifications ?: true
        fetchMembers(token)
    }

    fun fetchMembers(token: String?) {
        if (token == null) {
            return
        }
        viewModelScope.launch {
            try {
                SharedCacheService.getInstance(DatabaseDriverFactory(getApplication())).apiService()
                    .getChatMembers(
                        token,
                        conversation.groupType,
                        conversation.groupId
                    ).let {
                        members.value = it
                    }
            } catch (e: Exception) {
                e.printStackTrace()
            }
        }
    }

    fun updateMembership(token: String?) {
        if (token == null) {
            return
        }
        viewModelScope.launch {
            try {
                SharedCacheService.getInstance(DatabaseDriverFactory(getApplication())).apiService()
                    .putChatMembers(
                        token,
                        conversation.groupType,
                        conversation.groupId,
                        ChatMembershipUpload(
                            notifications.value ?: true
                        )
                    )
            } catch (e: Exception) {
                e.printStackTrace()
            }
        }
    }

}